source/Mlos.Python/mlos/Optimizers/ExperimentDesigner/UtilityFunctions/SeriesDifferenceUtilityFunction.py

```diff
@@ -0,0 +1,98 @@
+# Licensed under the MIT License.
+#
+import numpy as np
+import pandas as pd
+from scipy.stats import t
+from mlos.Logger import create_logger
+from mlos.Optimizers.OptimizationProblem import SeriesObjective
+from mlos.Optimizers.ExperimentDesigner.UtilityFunctions.UtilityFunction import UtilityFunction
+from mlos.Optimizers.RegressionModels.MultiObjectiveRegressionModel import MultiObjectiveRegressionModel
+from mlos.Optimizers.RegressionModels.Prediction import Prediction
+from mlos.Spaces import SimpleHypergrid, ContinuousDimension, CategoricalDimension, Point
+from mlos.Spaces.Configs.ComponentConfigStore import ComponentConfigStore
+from mlos.Tracer import trace
+
+
+series_difference_utility_function_config_store = ComponentConfigStore(
+    parameter_space=SimpleHypergrid(
+        name="series_difference_utility_function_config",
+        dimensions=[
+            # I might use some of these from ConfidenceBoundUtilityFunction... They seem pretty reasonable.
+            CategoricalDimension(name="utility_function_name", values=["lower_confidence_bound_on_improvement", "upper_confidence_bound_on_improvement"]),
+            ContinuousDimension(name="alpha", min=0.01, max=0.5)
+        ]
+    ),
+    default=Point(
+        utility_function_name="upper_confidence_bound_on_improvement",
+        alpha=0.01
+    )
+)
+
+
+class SeriesDifferenceUtilityFunction(UtilityFunction):
+    def __init__(self, function_config: Point, surrogate_model: MultiObjectiveRegressionModel, objective: SeriesObjective, logger=None):
+        if logger is None:
+            logger = create_logger(self.__class__.__name__)
+        self.logger = logger
+
+        self.config = function_config
+        self.series_objective = objective
+        self._sign = 1 if not self.series_objective.minimize else -1
+        self.surrogate_model: MultiObjectiveRegressionModel = surrogate_model
+
+    @trace()
+    def __call__(self, feature_values_pandas_frame):
+        # NOTE: THIS LOGIC IS DUPLICATED INTO SeriesPredictedValueUtilityFunction and SeriesDifferenceUtilityFunction.
+        # Later they will diverge as SeriesDifferenceUtilityFunction should require additional logic surrounding probability
+        # of improvement. Perhaps this part should be refactored to be shared though.
+        #
+        self.logger.debug(f"Computing utility values for {len(feature_values_pandas_frame.index)} points.")
+
+        predicted_value_col = Prediction.LegalColumnNames.PREDICTED_VALUE.value
+        predicted_value_var_col = Prediction.LegalColumnNames.PREDICTED_VALUE_VARIANCE.value
+        dof_col = Prediction.LegalColumnNames.PREDICTED_VALUE_DEGREES_OF_FREEDOM.value
+
+        del feature_values_pandas_frame[f"context_space.{self.series_objective.series_modulation_dimension.name}"]
+        series_vals_df = pd.DataFrame({
+            f"context_space.{self.series_objective.series_modulation_dimension.name}": self.series_objective.series_modulation_dimension.linspace()

```

Comment:
Modulation implies periodicity, no? How about series indexing dimension? Also, that dimension should really be a ContinuousDimension or maybe a DiscreteDimension. But according to all major religions, and philosophies, time is not Categorical :)

Reply:
Also, why use linspace here, instead of grabbing these values from target_series? That way, they will always be matched up...

Reply:
Yes but values in time are.... well not categorical but close. As I wrote earlier, if kusto only gives reports at specified intervals, it seems weird to assume that those intervals are going to be evenly spaced. 

Also, according to all major religions and philosophies, the number of instances between t=0s and t=1s is infinite, and yet this sadly cannot evaluate infinite points (yet..) Putting the values as discrete (and flexible) time intervals was my solution to this problem. Thoughts?